File: leetcode/two pointer/567. Permutation in String/js/main.js
// Time O(N)
// Space O(N)
const checkInclusion = (s1, s2) => {
  let map = {};

  for (let x of s1) map[x] = ~~map[x] + 1;

  let start = 0;
  let end = 0;
  let cnt = Object.keys(map).length;

  while (end < s2.length) {
    map[s2[end]]--;

    if (map[s2[end]] == 0) {
      cnt--;
    }

    end++;

    while (cnt == 0) {
      if (end - start == s1.length) {
        return true;
      }

      map[s2[start]]++;

      if (map[s2[start]] > 0) {
        cnt++;
      }

      start++;
    }
  }

  return false;
};

File: leetcode/two pointer/567. Permutation in String/py/main.py
import collections

# We need to find a minimum substring of s2 whose length equal s1
class Solution:
    # Time O(N)
    # Space O(N)
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        s1CountMap = collections.Counter(s1)
        start, end = 0, 0
        k = 0
        
        while end < len(s2):
            if s2[end] in s1CountMap:
                if s1CountMap[s2[end]] == 1:
                    k += 1
                s1CountMap[s2[end]] -= 1
            
            end += 1

            while k == len(s1CountMap):
              if end - start == len(s1):
                  return True
                
              if s2[start] in s1CountMap:
                  if s1CountMap[s2[start]] == 0:
                      k -= 1
                  s1CountMap[s2[start]] += 1
            
              start += 1
                
        return False
